src/modules/nix_shell.rs

```diff
@@ -3,6 +3,36 @@ use super::{Context, Module, ModuleConfig};
 use crate::configs::nix_shell::NixShellConfig;
 use crate::formatter::StringFormatter;
 
+enum NixShellType {
+    Pure,
+    Impure,
+    // We're in a Nix shell, but we don't know which type.
+    Unknown,
+}
+
+impl NixShellType {
+    fn detect_shell_type(context: &Context) -> Option<NixShellType> {
+        use NixShellType::*;
+
+        let shell_type = context.get_env("IN_NIX_SHELL");
+        let shell_type_format_from_env = match shell_type.as_deref() {
+            Some("pure") => Some(Pure),
+            Some("impure") => Some(Impure),
+            _ => None,
+        };
+
+        shell_type_format_from_env.or_else(|| Self::in_new_nix_shell(context).map(|_| Unknown))
+    }
+
+    // Hack to detect if we're in a `nix shell` (in constrast to a `nix-shell`).
+    // A better way to do this will be enabled by https://github.com/NixOS/nix/issues/6677.
+    fn in_new_nix_shell(context: &Context) -> Option<()> {
+        let path = context.get_env("PATH")?;
+
+        path.contains("/nix/store").then_some(())
```

Comment:
To reduce false positive, please use `env::split_paths` and check if any of the paths _starts_ with `/nix/store/`.

Reply:
Thank you, done in https://github.com/starship/starship/pull/4724/commits/5d369fb50adf8fc58e3929e41e338970f0489edd.